include/openssl/trace.h

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef OSSL_TRACE_H
+# define OSSL_TRACE_H
+
+# include <stdarg.h>
+
+# include <openssl/bio.h>
+
+# ifdef  __cplusplus
+extern "C" {
+# endif
+
+/*
+ * OSSL_tracer_fn is the type of the function that the application defines.
+ * It MUST return the number of bytes written, or 0 on error (in other words,
+ * it can never write zero bytes).
+ *
+ * The data to write will always be text, and may be several lines.
+ */
+typedef size_t (OSSL_tracer_fn)(const char *buf, size_t cnt, void *hookdata);
+
+/*
+ * The DEBUG and TRACE types or "channels" that the application can register
+ * a hook for.  The DEFAULT type works as a single fallback.
+ */
+# define OSSL_DEBUG_DEFAULT             0 /* The fallback */
+# define OSSL_DEBUG_INIT                1
+# define OSSL_DEBUG_SSL                 2
+# define OSSL_DEBUG_SSL_CIPHER          3
```

Comment:
Should these be named '*_TLS*' rather than '*_SSL*'?  Is it time that we shake off the SSL shackles?

(I could make aliases to make both variants available)